nvflare/private/fed/server/server_engine.py

```diff
@@ -220,23 +252,26 @@ def _start_runner_process(self, args, app_root, run_number, app_custom_folder, s
         for t in args.set:
             command_options += " " + t
         command = (
-            f"{sys.executable} -m nvflare.private.fed.app.server.runner_process -m "
-            + args.workspace
-            + " -s fed_server.json -r "
-            + app_root
-            + " -n "
-            + str(run_number)
-            + " -p "
-            + str(self.open_port)
-            + " --set"
-            + command_options
-            + " print_conf=True restore_snapshot="
-            + str(restore_snapshot)
+                f"{sys.executable} -m nvflare.private.fed.app.server.runner_process -m "
+                + args.workspace
+                + " -s fed_server.json -r " + app_root
+                + " -n " + str(run_number)
+                + " -p " + str(listen_port)
+                + " -c " + str(open_ports[0])
+                + " --set" + command_options + " print_conf=True restore_snapshot=" + str(restore_snapshot)
         )
         # use os.setsid to create new process group ID
 
         # command = f"{sys.executable} -m nvflare.private.fed.app.server.worker_process"
         process = subprocess.Popen(shlex.split(command, " "), preexec_fn=os.setsid, env=new_env)
+
+        threading.Thread(target=self.wait_for_complete, args=[run_number]).start()
+
+        with self.lock:
+            self.run_processes[run_number] = {"_port": listen_port,
+                                              "_conn": None,
+                                              "_child_process": process
```

Comment:
it might be good to specify these strings as CONSTANTS up front.

For example, `__PORT = "_port"`

And when referring to these, we can use the var name __PORT

Reply:
done